File: lab10.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


typedef struct Node{
    char *value;
    struct LinkedList *neighbours;
}Node;

typedef struct LinkedListNode{
    char *value;
    struct LinkedListNode *next;
}LinkedListNode;

typedef struct LinkedList{
    LinkedListNode *head;
}LinkedList;

void create_node(Node *node, char *value){
    node->value = (char *)malloc(strlen(value) + 1);
    strcpy(node->value, value);
    node->neighbours = malloc(sizeof(LinkedList));
    node->neighbours->head = malloc(sizeof(LinkedListNode));
}

void create_LinkedListNode(LinkedListNode *node, char *value){
    node->value = (char *)malloc(strlen(value) + 1);
    strcpy(node->value, value);
    node->next = NULL;
}

void create_LinkedList(LinkedList *list){
    list->head = (LinkedListNode *)malloc(sizeof(LinkedListNode));
}

void add(LinkedList *list, char *value){
    LinkedListNode *new_node = (LinkedListNode *)malloc(sizeof(LinkedListNode));
    create_LinkedListNode(new_node, value);
    if (list->head->next == NULL){
        create_LinkedListNode(list->head, value);
    }
    else{
        LinkedListNode *current = list->head;
        while(current->next != NULL){
            current = current->next;
        }
        create_LinkedListNode(current->next, value);
    }
}

int are_nodes_linked(Node node1, Node node2)
{
    LinkedListNode *current = node1.neighbours->head;
    while(current != NULL){
        if(strcmp(current->value, node2.value) == 0){
            // If present in the neighbour list, return 1
            return 1;
        }
        current = current->next;
    }
    return 0; // If not present in neighbour list, return 0
}

Node *get_node_by_str(Node **nodes, char *node_str)
{
    for(int i = 0; i < sizeof(nodes)/sizeof(nodes[0]); i++){
       if(strcmp(nodes[i]->value, node_str) == 0)
       {
            return nodes[i];
       }
    }
    return NULL;
}

int are_airports_linked(Node **airports, char *airport1, char *airport2){
// Return 1, the airports are linked
// Return -1, the airports are not linked
{
    Node *airport1_node = get_node_by_str(airports, airport1);
    Node *airport2_node = get_node_by_str(airports, airport2);
    if(airport1_node == NULL || airport2_node == NULL){
        return 0;
    }
    return are_nodes_linked(*airport1_node, *airport2_node);
}
}

Node **make_airport_graph(){
    Node *yyz = malloc(sizeof(Node));
    Node *yvr = malloc(sizeof(Node));
    Node *yul = malloc(sizeof(Node));
    Node *whitehorse = malloc(sizeof(Node));
    create_node(yyz, "YYZ");
    create_node(yvr, "YVR");
    create_node(yul, "YUL");
    create_node(whitehorse, "Whitehorse");

    add(yyz->neighbours, "YYZ");
    add(yyz->neighbours, "YUL");
    add(yvr->neighbours, "YYZ");
    add(yvr->neighbours, "YUL");
    add(yvr->neighbours, "Whitehorse");
    add(yul->neighbours, "YYZ");
    add(yul->neighbours, "YVR");
    add(whitehorse->neighbours, "YVR");

    Node **airports = malloc(4 * sizeof(Node *));
    airports[0] = yyz;
    airports[1] = yvr; 
    airports[2] = yul;
    airports[3] = whitehorse;
    return airports;
}

int main(){
    Node **airports = make_airport_graph();
    if(are_airports_linked(airports, "YYZ", "YUL") == 1){
        printf("Toronto (YYZ) is connected to Montreal (YUL)\n");
    }
    else{
        printf("Toronto (YYZ) is not connected to Montreal (YUL)\n");
    }   
}

File: lab10.py
import numpy as np

class Node:
    def __init__(self, value):
        self.value = value
        self.neighbours = LinkedList()

class LinkedListNode:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def add(self, value):
        new_node = LinkedListNode(value)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
    
    def __iter__(self):
        current = self.head
        yield current.value
        current = current.next
    
def are_nodes_linked(node1, node2):
    return node2.value in node1.neighbours

def get_node_by_str(nodes, node_str):
    return nodes.get(node_str, None)

def are_airports_linked(airports, airport1, airport2):
    airport1 = get_node_by_str(airports, airport1)
    airport2 = get_node_by_str(airports, airport2)
    if airport1 is None or airport2 is None:
        return False
    return are_nodes_linked(airport1, airport2)

def convert_to_adj_matrix(airports):
    num_airports = len(airports) # Obtain the length of the airport dictionary
    adj_matrix = np.zeros((num_airports, num_airports), dtype=int) # Create the adjacency matrix with integer zeros
    airport_callsigns = sorted(list(airports.keys()))
    airport_callsigns_idx = {airport_callsign: i for i, airport_callsign in enumerate(airport_callsigns)}

    for airport_callsign, airport in airports.items():
        for neighbour_node in airport.neighbours:
            neighbour_callsign = neighbour_node.value
            adj_matrix[airport_callsigns_idx[airport_callsign], airport_callsigns_idx[neighbour_callsign]] = 1

    return adj_matrix, airport_callsigns_idx

def make_airport_graph1():
    # Nodes for graph making
    yyz = Node("YYZ")
    yvr = Node("YVR")
    yul = Node("YUL")
    whitehorse = Node("Whitehorse")

    # Access the neighbours list for the nodes and add airports
    yyz.neighbours.add(yvr)
    yyz.neighbours.add(yul)
    yvr.neighbours.add(yyz)
    yvr.neighbours.add(yul)
    yvr.neighbours.add(whitehorse)
    yul.neighbours.add(yyz)
    yul.neighbours.add(yvr)
    whitehorse.neighbours.add(yvr)

    airport_dict = {}
    for airport in [yyz, yvr, yul, whitehorse]:
        airport_dict[airport.value] = airport

    return airport_dict

if __name__ == "__main__":
    airports = make_airport_graph1()
    print("Toronto (YYZ) is connected to Montreal (YUL)?", are_airports_linked(airports, "YYZ", "YUL"))